server/src/api/http/oidc.py

```diff
@@ -0,0 +1,257 @@
+        """Whether to use PKCE (Proof Key for Code Exchange) during authentication"""
+
+
+@dataclass
+class ExchangeData:
+    """Standardized data returned after exchanging code for token"""
+    username: str
+    email: str
+
+class OIDCAuth:
+    """Handles OIDC authentication logic
+    
+    Provides methods to get authorization URLs, exchange codes for tokens, and manages multiple OIDC providers.
+    """
+    # Set to the minimum time so that we always initialize on first call
+    last_update: datetime = datetime.min.replace(tzinfo=UTC)
+
+    def __init__(self):
+        #self.last_update = datetime.min
+        # For each configured OIDC provider, create an OidcConfig
+        self.code_cache = CodeCache()
+        self.providers = dict[str, OidcServerConfig]()
+
+    def get_provider(self, provider_id: str) -> OidcServerConfig | None:
+        """Get the OIDC provider configuration by ID
+        
+        Also refreshes the providers from the config if needed."""
+        # check the config and remove any providers that are no longer present
+        # Generate the redirect URI based on the server config
+        # but only if the config has changed since last update
+        if cfg_last_update() > self.last_update:
+            self.last_update = cfg_last_update()
+            logger.debug("Refreshing OIDC providers from config")
+            redirect_uri = (cfg().general.client_url or "") + "/api/oidc/callback/"
+
+            # Mark all providers as not refreshed first
+            for provider_id in list(self.providers.keys()):
+                self.providers[provider_id].refreshed = False
+            
+            # Now refresh from the config
+            for oidc_cfg in cfg().oidc:
+                provider: OidcServerConfig
+                if oidc_cfg.provider_id not in self.providers:
+                    provider = OidcServerConfig(oidc_cfg, redirect_uri)
+                    self.providers[oidc_cfg.provider_id] = provider
+                else:
+                    provider = self.providers[oidc_cfg.provider_id]
+                provider.refreshed = True
+            
+            # Remove any providers that were not refreshed
+            for provider_id in list(self.providers.keys()):
+                if not self.providers[provider_id].refreshed:
+                    del self.providers[provider_id]
+
+        return self.providers.get(provider_id)
+
+    def get_providers(self) -> list[dict] | None:
+        """Get a list of configured OIDC providers
+        
+            Returns a list of dictionaries with provider display names and IDs."""
+        # This is often the first call, so we will attempt to get a None provider
+        self.get_provider("") # Trigger load if needed
+        return [
+            {
+                "display_name": provider.display_name,
+                "provider_id": provider.provider_id,
+            }
+            for provider in self.providers.values()
```

Comment:
Given my comment above on `get_provider`, I would just drop the `self.providers` intermediate state and always just iterate over the config values here.

That also means you no longer need the funky "" load call